tensornetwork/backends/backend_test.py

```diff
@@ -56,18 +56,3 @@ def test_config_tensorflow_missing_can_import_config():
   import tensornetwork.config
   with pytest.raises(ImportError):
     import tensorflow as tf
-
-
-@pytest.mark.usefixtures('no_backend_dependency')
```

Comment:
Why did you delete this?

Reply:
It doesn't really test anything, as far as I can tell. All it's testing is that the mock import, defined in the same test file, raises an import error. So rather than fix the test for the other changes, it seems more sensible to just delete it

Reply:
Actually, the same can be said for all the other tests in that same file.

Reply:
Then we need to fix them.

Reply:
The issue is we have no good tests for ensuring that when certain backends aren't installed, we can still use TensorNetwork fine for the other installed backends. These mock imports should have fixed this issue

Reply:
Can you something like this? It will ensure we can still use TensornNetwork, but will also ensure we raise a smart error if one of the backends is missing.

```python
import tensornetwork as tn
tn.TensorNetwork("numpy")
with pytest.raises(ImportError):
  tn.TensorNetwork("tensorflow")
```

Reply:
OK, updated import tests. Note that I've moved the imports so that an import error is thrown as soon as the backend module is imported, rather than waiting until the backend class is instantiated.

Reply:
Dont agree with this? The import should NOT raise an error? Also, for me, at least very naively, the tests do work (i.e. when I put an `import torch` in the init of the package, the tests fail. Not sure all is covered, but something is?

Reply:
Import errors should be raised on import, not (generally speaking) when you instantiate some class.

Reply:
I agree in principle. But now we need to make sure none of the modules are imported by main package and this is I guess where the internally broken issue mentioned by @Thenerdstation comes in (just trying to understand that one)?

Reply:
The main issue is that our internal build doesn't like dynamic imports. I hit this error when I tried using your branch internally.

```
tensornetwork/backends/backend_factory.py in get_backend(name, dtype)
     35   module_name, class_name = _BACKENDS[name]
     36   package_name = 'tensornetwork.backends'
---> 37   backend_module = importlib.import_module(module_name, package_name)
     38   backend_class = getattr(backend_module, class_name)
     39   dtypes = getattr(backend_module, 'supported_dtypes')


ModuleNotFoundError: No module named 'tensornetwork.backends.numpy'
```

Reply:
@gecrooks  
> Actually, the same can be said for all the other tests in that same file.
Is there still an additional issue now with the tests?

Reply:
No, I think I now see what they are trying to do.

Reply:
OK, I don't need to use importlib. Can use conditional imports instead. Just less elegant.

Reply:
@gecrooks 

> No, I think I now see what they are trying to do.

Thanks for confirming. They are a bit hacky & fragile though as you really need to cleanup all traces of all imported submodules... If you know of a more elegant way, I'd be happy to learn!

Reply:
Still a no
```
tensornetwork/backends/backend_factory.py in get_backend(name, dtype)
     27     dtypes = tensorflow_backend.supported_dtypes
     28   elif name == "numpy":
---> 29     from .numpy import numpy_backend
     30     backend_class = numpy_backend.NumPyBackend
     31     dtypes = numpy_backend.supported_dtypes

ModuleNotFoundError: No module named 'tensornetwork.backends.numpy'
```

If the imports aren't at the top, they don't get included internally.

@gecrooks, just let me take the lead on this. There are a lot of subtle issues that need to be fixed.